editorconfigCheck fails due to incompatible version of antlr4-runtime

gradlew editorConfigCheck

..

> Task :app:editorconfigCheck FAILED
ANTLR Tool version 4.7 used for code generation does not match the current runtime version 4.5.3

I have an android application with the following root build.gradle:

```
buildscript {
  repositories {
    google()
    jcenter()
  }
  dependencies {
    classpath 'com.android.tools.build:gradle:3.1.4'
 }
}

plugins {
  id 'org.ec4j.editorconfig' version '0.0.2'
}
```

The app build.gradle is nothing special:
```
apply plugin: 'com.android.application'

android {
  compileSdkVersion 26
  defaultConfig {
    applicationId "some.app.id"
    minSdkVersion 19
    targetSdkVersion 26
    versionCode 1
    versionName "1.0"
    testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
  }
  buildTypes {
    release {
      minifyEnabled false
      proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
    }
  }
}

configurations.all {
  // Don't cache changing modules at all.
  resolutionStrategy.cacheChangingModulesFor 0, 'seconds'
}

dependencies {
  implementation fileTree(dir: 'libs', include: ['*.jar'])
  implementation 'com.android.support:appcompat-v7:26.1.0'
}
```
when I run the command 'gradlew editorConfigCheck' I get the following output:

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':editorconfigCheck'.
> Could not initialize class org.ec4j.maven.linters.xml.XmlParser
```

Here's a simple reproduction scenario:
clone a google sample project
```git clone https://github.com/googlesamples/android-CardView.git```

edit Application\build.gradle:
```
buildscript {
    repositories {
        jcenter()
        google()
    }

    dependencies {
        classpath 'com.android.tools.build:gradle:3.0.1'
    }
}

plugins {
  id 'org.ec4j.editorconfig' version '0.0.2'
}

apply plugin: 'com.android.application'
...
```
Put .editorconfig file in root dir:
```
[*]
charset=utf-8
end_of_line=lf
insert_final_newline=false
indent_style=space
indent_size=2

[{*.sht,*.html,*.shtm,*.shtml,*.htm}]
indent_style=space
indent_size=4

[{*.mod,*.dtd,*.ent,*.elt}]
indent_style=space
indent_size=4

[{*.jhm,*.xslt,*.xul,*.rng,*.xsl,*.xsd,*.ant,*.tld,*.fxml,*.jrxml,*.xml,*.jnlp,*.wsdl}]
indent_style=space
indent_size=4

[{org.codehaus.groovy.runtime.ExtensionModule,*.properties}]
indent_style=space
indent_size=4
```

run gradlew editorconfigCheck:
```
> Configure project :Application
Configuration 'compile' in project ':Application' is deprecated. Use 'implementation' instead.
The CompileOptions.bootClasspath property has been deprecated and is scheduled to be removed in Gradle 5.0. Please use the CompileOptions.bootstrapClasspath property instead.

> Task :Application:editorconfigCheck FAILED
ANTLR Tool version 4.7 used for code generation does not match the current runtime version 4.5.3ANTLR Runtime version 4.7 used for parser compilation does not match the current runtime version 4.5.3

FAILURE: Build failed with an exception.
```

--- a/build.gradle
+++ b/build.gradle
@@ -38,7 +38,8 @@ repositories {
 dependencies {
     compile gradleApi()
     compile localGroovy()
-    compile 'org.ec4j.maven:ec4j-linters:0.0.5'
+    compile 'org.ec4j.maven:ec4j-lint-api:0.0.7' // Keep in sync with org.ec4j.gradle.EditorconfigGradlePlugin.LINTERS_VERSION */
+
     testCompile 'junit:junit:4.12'
     testCompile gradleTestKit()
     testCompile 'org.slf4j:slf4j-api:1.7.25'
@@ -68,6 +69,13 @@ pluginBundle {
     }
 }
 
+test {
+    testLogging.showStandardStreams = true
+    testLogging {
+        exceptionFormat = 'full'
+    }
+}
+
 task javadocJar(type: Jar) {
     classifier = 'javadoc'
     from javadoc
@@ -90,7 +98,9 @@ license {
     ext.year = '2018'
     excludes ([
         "**/*.adoc",
+        "**/.classpath",
         "**/.gitkeep",
+        "**/.project",
         ".mvn/wrapper/",
         "LICENSE",
         "gradlew",
@@ -99,6 +109,7 @@ license {
         "src/test/projects/**/*.txt"
     ])
     mapping {
+        editorconfig = 'SCRIPT_STYLE'
         java = 'SLASHSTAR_STYLE'
     }
 }

--- a/gradle/wrapper/gradle-wrapper.properties
+++ b/gradle/wrapper/gradle-wrapper.properties
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-3.5-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.10.2-bin.zip

--- a/src/main/java/org/ec4j/gradle/AbstractEditorconfigTask.java
+++ b/src/main/java/org/ec4j/gradle/AbstractEditorconfigTask.java
@@ -17,33 +17,28 @@
 package org.ec4j.gradle;
 
 import java.io.File;
-import java.io.IOException;
 import java.nio.charset.Charset;
-import java.nio.file.Path;
 import java.util.LinkedHashSet;
-import java.util.List;
+import java.util.Map.Entry;
 import java.util.Set;
 
-import org.ec4j.core.Cache.Caches;
-import org.ec4j.core.Resource.Resources;
-import org.ec4j.core.ResourceProperties;
-import org.ec4j.core.ResourcePropertiesService;
-import org.ec4j.core.model.PropertyType;
+import org.ec4j.gradle.CollectingLogger.LogMessages;
+import org.ec4j.gradle.runtime.EditorconfigInvoker;
 import org.ec4j.maven.lint.api.Constants;
-import org.ec4j.maven.lint.api.EditableResource;
-import org.ec4j.maven.lint.api.FormatException;
-import org.ec4j.maven.lint.api.Linter;
-import org.ec4j.maven.lint.api.LinterRegistry;
-import org.ec4j.maven.lint.api.Resource;
-import org.ec4j.maven.lint.api.ViolationHandler;
 import org.gradle.api.Action;
 import org.gradle.api.DefaultTask;
 import org.gradle.api.GradleException;
 import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
 import org.gradle.api.file.ConfigurableFileTree;
 import org.gradle.api.file.FileTree;
+import org.gradle.api.logging.LogLevel;
 import org.gradle.api.logging.Logger;
 import org.gradle.api.tasks.TaskAction;
+import org.gradle.workers.IsolationMode;
+import org.gradle.workers.WorkerConfiguration;
+import org.gradle.workers.WorkerExecutionException;
+import org.gradle.workers.WorkerExecutor;
 
 /**
  * A base for {@link EditorconfigCheckTask} and {@link EditorconfigFormatTask}.
@@ -52,41 +47,16 @@ import org.gradle.api.tasks.TaskAction;
  */
 public abstract class AbstractEditorconfigTask extends DefaultTask {
 
-    private static LinterRegistry buildLinterRegistry(EditorconfigExtension ext) {
-
-        final LinterRegistry.Builder linterRegistryBuilder = LinterRegistry.builder();
-
-        final ClassLoader cl = ext.getClass().getClassLoader();
-        if (ext.isAddLintersFromClassPath()) {
-            linterRegistryBuilder.scan(cl);
-        }
-
-        final List<LinterConfig> linters = ext.getLinters();
-        if (linters != null && !linters.isEmpty()) {
-            for (LinterConfig linter : linters) {
-                if (linter.isEnabled()) {
-                    linterRegistryBuilder.entry(linter.getId(), linter.getClassName(), cl, linter.getIncludes(),
-                            linter.getExcludes(), linter.isUseDefaultIncludesAndExcludes());
-                } else {
-                    linterRegistryBuilder.removeEntry(linter.getId());
-                }
-            }
-        }
-
-        return linterRegistryBuilder.build();
-
-    }
+    private final WorkerExecutor workerExecutor;
 
     /**
      * {@link FileTree} scanning boilerplate.
      *
-     * @param project
-     *            the project being built
-     * @param editorconfigExtension
-     *            the config, cannot be null
-     * @return a {@link Set} of included files
+     * @param project the project being built
+     * @param editorconfigExtension the config, cannot be null
+     * @return a {@link Set} of absolute paths of included files
      */
-    private static Set<File> scanIncludedFiles(Project project, final EditorconfigExtension editorconfigExtension) {
+    private static Set<String> scanIncludedFiles(Project project, final EditorconfigExtension editorconfigExtension) {
         FileTree tree = project.fileTree(project.getProjectDir(), new Action<ConfigurableFileTree>() {
             @Override
             public void execute(ConfigurableFileTree tree) {
@@ -99,26 +69,20 @@ public abstract class AbstractEditorconfigTask extends DefaultTask {
                 tree.exclude(excls);
             }
         });
-        return tree.getFiles();
+        final Set<String> result = new LinkedHashSet<>();
+        for (File file : tree.getFiles()) {
+            result.add(file.getAbsolutePath());
+        }
+        return result;
     }
 
-    /** Rhe {@link EditorconfigExtension} initialized in {@link #perform()} */
+    /** The {@link EditorconfigExtension} initialized in {@link #perform()} */
     protected EditorconfigExtension editorconfigExtension;
 
-    protected abstract ViolationHandler createHandler();
-
-    /**
-     * Create a new {@link Resource} suitable for the current task.
-     *
-     * @param absFile
-     *            the {@link Path} to create a {@link Resource} for. Must be absolute.
-     * @param relFile
-     *            the {@link Path} to create a {@link Resource} for. Must be relative to {@code project.getProjectDir()}.
-     * @param encoding
-     *            the encoding of the resulting {@link Resource}
-     * @return a new {@link Resource} or a new {@link EditableResource}
-     */
-    protected abstract Resource createResource(Path absFile, Path relFile, Charset encoding);
+    protected AbstractEditorconfigTask(WorkerExecutor workerExecutor) {
+        super();
+        this.workerExecutor = workerExecutor;
+    }
 
     /**
      * Performs this task.
@@ -141,56 +105,67 @@ public abstract class AbstractEditorconfigTask extends DefaultTask {
         } else {
             charset = Charset.forName(editorconfigExtension.getEncoding());
         }
-        final Path basedirPath = project.getProjectDir().toPath();
+        final String basedirPath = project.getProjectDir().toPath().toString();
 
-        LinterRegistry linterRegistry = buildLinterRegistry(editorconfigExtension);
-        final Set<File> includedFiles = scanIncludedFiles(project, editorconfigExtension);
+        final Set<String> includedPaths = scanIncludedFiles(project, editorconfigExtension);
+
+        final Configuration classpath = project.getConfigurations().getAt(EditorconfigGradlePlugin.CONFIGURATTION_NAME);
+
+        workerExecutor.submit(EditorconfigInvoker.class, new Action<WorkerConfiguration>() {
+            @Override
+            public void execute(WorkerConfiguration config) {
+                config.setIsolationMode(IsolationMode.CLASSLOADER);
+                config.params(AbstractEditorconfigTask.this.getClass().getName(), includedPaths, basedirPath,
+                        charset.name(), editorconfigExtension.isFailOnFormatViolation(),
+                        editorconfigExtension.isBackup(), editorconfigExtension.getBackupSuffix(),
+                        editorconfigExtension.isAddLintersFromClassPath(), editorconfigExtension.getLinters(),
+                        editorconfigExtension.isFailOnNoMatchingProperties());
+                config.classpath(classpath);
+            }
+        });
 
         try {
-            final ViolationHandler handler = createHandler();
+            workerExecutor.await();
+        } catch (WorkerExecutionException e) {
 
-            final ResourcePropertiesService resourcePropertiesService = ResourcePropertiesService.builder() //
-                    .cache(Caches.permanent()) //
-                    .build();
-            handler.startFiles();
-            boolean propertyMatched = false;
-            for (File includedFile : includedFiles) {
-                final Path absFile = includedFile.getAbsoluteFile().toPath();
-                final Path file = basedirPath.relativize(absFile);
-                log.debug("Processing file '{}'", file);
-                final ResourceProperties editorConfigProperties = resourcePropertiesService
-                        .queryProperties(Resources.ofPath(absFile, charset));
-                if (!editorConfigProperties.getProperties().isEmpty()) {
-                    propertyMatched = true;
-                    final Charset useEncoding = Charset.forName(editorConfigProperties.getValue(PropertyType.charset,
-                            editorconfigExtension.getEncoding(), true));
-                    final Resource resource = createResource(absFile, file, useEncoding);
-                    final List<Linter> filteredLinters = linterRegistry.filter(file);
-                    ViolationHandler.ReturnState state = ViolationHandler.ReturnState.RECHECK;
-                    while (state != ViolationHandler.ReturnState.FINISHED) {
-                        for (Linter linter : filteredLinters) {
-                            if (log.isDebugEnabled()) {
-                                log.debug("Processing file '{}' using linter {}", file, linter.getClass().getName());
-                            }
-                            handler.startFile(resource);
-                            linter.process(resource, editorConfigProperties, handler);
+            /* A megahack to pass data from the classpath-isolated WorkerExecutor.
+             * Found no better way than to use some Exception defined in JDK
+             * and smuggle the data through its message field as a string.
+             * The serialization format is defined in CollectingLogger.
+             * Note that even the log messages need to be passed like that because
+             * loggers instantiated by the isolated class loader do not obey
+             * the log level set via Gradle CLI.
+             * @ppalaga is open for suggestions to improve this :)
+             */
+
+            Throwable t = e;
+            while (t.getCause() != null) {
+                t = t.getCause();
+                if (t instanceof RuntimeException) {
+                    final String msg = t.getMessage();
+                    if (msg != null && msg.startsWith(EditorconfigInvoker.FORMAT_EXCEPTION_PREFIX)) {
+                        final LogMessages logData = CollectingLogger
+                                .deserialize(EditorconfigInvoker.FORMAT_EXCEPTION_PREFIX.length(), msg);
+                        for (Entry<String, String> logMessage : logData.getMessages()) {
+                            final String key = logMessage.getKey();
+                            /* There is no TRACE in org.gradle.api.logging.LogLevel */
+                            final LogLevel logLevel = org.ec4j.maven.lint.api.Logger.LogLevel.TRACE.name().equals(key)
+                                    ? LogLevel.DEBUG
+                                    : LogLevel.valueOf(key);
+                            log.log(logLevel, logMessage.getValue());
                         }
-                        state = handler.endFile();
+                        final String failureMessage = logData.getFailureMessage();
+                        if (failureMessage != null) {
+                            throw new GradleException(failureMessage);
+                        } else {
+                            return;
+                        }
                     }
                 }
             }
-            if (!propertyMatched) {
-                if (editorconfigExtension.isFailOnNoMatchingProperties()) {
-                    log.error("No .editorconfig properties applicable for files under '{}'", basedirPath);
-                } else {
-                    log.warn("No .editorconfig properties applicable for files under '{}'", basedirPath);
-                }
-            }
-            handler.endFiles();
-        } catch (IOException e) {
-            throw new GradleException(e.getMessage(), e);
-        } catch (FormatException e) {
-            throw new GradleException("\n\n" + e.getMessage() + "\n\n", e);
+            throw e;
         }
+
     }
+
 }

--- a/src/main/java/org/ec4j/gradle/CollectingLogger.java
+++ b/src/main/java/org/ec4j/gradle/CollectingLogger.java
@@ -1,0 +1,221 @@
+/*
+ * Copyright (c) 2018 EditorConfig Gradle Plugin
+ * project contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ec4j.gradle;
+
+import java.io.IOException;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.ec4j.gradle.runtime.EditorconfigInvoker;
+import org.ec4j.maven.lint.api.Logger;
+
+/**
+ * A {@link Logger} that collects the log messages in a {@link #messages} {@link List}. This class exists because the
+ * SLF4J loggers instantiated by the isolated class loader inside {@link EditorconfigInvoker} do not obey the log level
+ * set via Gradle CLI. Thus we collect the messages in a {@link List}, serialize them to {@link String} (see
+ * {@link #serialize(Appendable)}), send them via and exception message to the caller context where we
+ * {@link #deserialize(int, String)} them to {@link LogMessages}.
+ *
+ * @author <a href="https://github.com/ppalaga">Peter Palaga</a>
+ * @since 0.0.3
+ */
+public class CollectingLogger extends Logger.AbstractLogger {
+
+    /**
+     * An immutable {@link List} of log {@link #messages}, optionally with a {@link #failureMessage}.
+     */
+    public static class LogMessages {
+
+        private final String failureMessage;
+        private final List<Map.Entry<String, String>> messages;
+
+        LogMessages(List<Entry<String, String>> messages, String failureMessage) {
+            super();
+            this.messages = messages;
+            this.failureMessage = failureMessage;
+        }
+
+        /**
+         * @return the failure message, can be {@code null} if the underlying task did not fail.
+         */
+        public String getFailureMessage() {
+            return failureMessage;
+        }
+
+        /**
+         * @return an unmodifiable {@link List} of log message {@link Entry}s. Keys are {@link LogLevel}s and values are
+         *         the messages.
+         */
+        public List<Map.Entry<String, String>> getMessages() {
+            return messages;
+        }
+    }
+
+    /**
+     * Parses the output of {@link CollectingLogger#serialize(Appendable)} to {@link LogMessages}.
+     */
+    static class Parser {
+
+        private static int unescape(int offset, String in, StringBuilder out) {
+            int i = offset;
+            while (i < in.length()) {
+                char ch = in.charAt(i++);
+                switch (ch) {
+                case '\n':
+                    return i;
+                case '\\':
+                    char ch1 = in.charAt(i++);
+                    switch (ch1) {
+                    case 'n':
+                        out.append('\n');
+                        break;
+                    case 'r':
+                        out.append('\r');
+                        break;
+                    case '\\':
+                        out.append('\\');
+                        break;
+                    default:
+                        throw new IllegalStateException("Unecpxpected escape sequence [\\" + ch1 + "]");
+                    }
+                    break;
+                default:
+                    out.append(ch);
+                }
+            }
+            return i;
+        }
+
+        private String failureMessage = null;
+
+        private List<Map.Entry<String, String>> messages = new ArrayList<>();
+
+        Parser() {
+        }
+
+        private int entry(int offset, String in) {
+            final int spaceOffset = in.indexOf(' ', offset);
+            if (spaceOffset >= 0) {
+                final String key = in.substring(offset, spaceOffset);
+                final StringBuilder value = new StringBuilder();
+                final int lfOffset = unescape(spaceOffset + 1, in, value);
+                if (FAILURE.equals(key)) {
+                    if (this.failureMessage != null) {
+                        throw new IllegalStateException("Only one " + FAILURE + " expected");
+                    }
+                    this.failureMessage = value.toString();
+                } else {
+                    messages.add(new AbstractMap.SimpleImmutableEntry<>(key, value.toString()));
+                }
+                return lfOffset;
+            } else {
+                throw new IllegalStateException("Expected a key");
+            }
+        }
+
+        LogMessages parse(int offset, String in) {
+            while ((offset = entry(offset, in)) < in.length()) {
+                /* empty */
+            }
+            List<Entry<String, String>> m = this.messages;
+            this.messages = null;
+            return new LogMessages(Collections.unmodifiableList(m), failureMessage);
+        }
+    }
+
+    private static final String FAILURE = "FAILURE";
+
+    /**
+     * Parse the {@code input} starting at {@code offset} to {@link LogMessages}.
+     *
+     * @param offset the zero based index to start at the parsing in {@code input}
+     * @param input the string to parse
+     * @return new {@link LogMessages}
+     */
+    public static LogMessages deserialize(int offset, String input) {
+        return new Parser().parse(offset, input);
+    }
+
+    private static void escape(String string, Appendable out) throws IOException {
+        for (int i = 0; i < string.length(); i++) {
+            char ch = string.charAt(i);
+            switch (ch) {
+            case '\n':
+                out.append("\\n");
+                break;
+            case '\r':
+                out.append("\\r");
+                break;
+            case '\\':
+                out.append("\\\\");
+                break;
+            default:
+                out.append(ch);
+            }
+        }
+    }
+
+    private volatile String failureMessage;
+
+    private final List<Map.Entry<String, String>> messages = new ArrayList<>();
+
+    public CollectingLogger(LogLevel level) {
+        super(level);
+    }
+
+    /**
+     * Sets the {@link #failureMessage}
+     *
+     * @param failureMessage the failure message
+     */
+    public void fail(String failureMessage) {
+        if (this.failureMessage != null) {
+            throw new IllegalStateException();
+        }
+        this.failureMessage = failureMessage;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void log(LogLevel level, String string, Object... args) {
+        messages.add(new AbstractMap.SimpleImmutableEntry<String, String>(level.name(),
+                Slf4jFormatter.format(string, args)));
+    }
+
+    /**
+     * Append the {@link #messages} and {@link #failureMessage} to the given {@code output}.
+     *
+     * @param output the {@link Appendable} to append to
+     * @throws IOException thrown from {@link Appendable#append(CharSequence)}
+     */
+    public void serialize(Appendable output) throws IOException {
+        for (Map.Entry<String, String> entry : messages) {
+            output.append(entry.getKey()).append(' ');
+            escape(entry.getValue(), output);
+            output.append('\n');
+        }
+        if (failureMessage != null) {
+            output.append(FAILURE).append(' ');
+            escape(failureMessage, output);
+        }
+    }
+
+}

--- a/src/main/java/org/ec4j/gradle/EditorconfigCheckTask.java
+++ b/src/main/java/org/ec4j/gradle/EditorconfigCheckTask.java
@@ -16,12 +16,9 @@
  */
 package org.ec4j.gradle;
 
-import java.nio.charset.Charset;
-import java.nio.file.Path;
+import javax.inject.Inject;
 
-import org.ec4j.maven.lint.api.Resource;
-import org.ec4j.maven.lint.api.ViolationCollector;
-import org.ec4j.maven.lint.api.ViolationHandler;
+import org.gradle.workers.WorkerExecutor;
 
 /**
  * Checks whether files are formatted according to rules defined in {@code .editorconfig} files. If fomat violations are
@@ -35,16 +32,9 @@ public class EditorconfigCheckTask extends AbstractEditorconfigTask {
 
     public static final String NAME = "editorconfigCheck";
 
-    /** {@inheritDoc} */
-    @Override
-    protected ViolationHandler createHandler() {
-        return new ViolationCollector(editorconfigExtension.isFailOnFormatViolation(), "./gradlew editorconfigFormat");
-    }
-
-    /** {@inheritDoc} */
-    @Override
-    protected Resource createResource(Path absFile, Path relFile, Charset encoding) {
-        return new Resource(absFile, relFile, encoding);
+    @Inject
+    public EditorconfigCheckTask(WorkerExecutor workerExecutor) {
+        super(workerExecutor);
     }
 
 }

--- a/src/main/java/org/ec4j/gradle/EditorconfigFormatTask.java
+++ b/src/main/java/org/ec4j/gradle/EditorconfigFormatTask.java
@@ -16,13 +16,9 @@
  */
 package org.ec4j.gradle;
 
-import java.nio.charset.Charset;
-import java.nio.file.Path;
+import javax.inject.Inject;
 
-import org.ec4j.maven.lint.api.EditableResource;
-import org.ec4j.maven.lint.api.FormattingHandler;
-import org.ec4j.maven.lint.api.Resource;
-import org.ec4j.maven.lint.api.ViolationHandler;
+import org.gradle.workers.WorkerExecutor;
 
 /**
  * Formats a set of files so that they comply with rules defined in {@code .editorconfig} files.
@@ -33,15 +29,9 @@ import org.ec4j.maven.lint.api.ViolationHandler;
 public class EditorconfigFormatTask extends AbstractEditorconfigTask {
     public static final String NAME = "editorconfigFormat";
 
-    /** {@inheritDoc} */
-    @Override
-    protected ViolationHandler createHandler() {
-        return new FormattingHandler(editorconfigExtension.isBackup(), editorconfigExtension.getBackupSuffix());
+    @Inject
+    public EditorconfigFormatTask(WorkerExecutor workerExecutor) {
+        super(workerExecutor);
     }
 
-    /** {@inheritDoc} */
-    @Override
-    protected Resource createResource(Path absFile, Path relFile, Charset encoding) {
-        return new EditableResource(absFile, relFile, encoding);
-    }
 }

--- a/src/main/java/org/ec4j/gradle/EditorconfigGradlePlugin.java
+++ b/src/main/java/org/ec4j/gradle/EditorconfigGradlePlugin.java
@@ -18,6 +18,7 @@ package org.ec4j.gradle;
 
 import org.gradle.api.Plugin;
 import org.gradle.api.Project;
+import org.gradle.api.artifacts.dsl.DependencyHandler;
 
 /**
  * Adds {@link EditorconfigExtension}, {@link EditorconfigCheckTask} and {@link EditorconfigFormatTask}.
@@ -25,11 +26,19 @@ import org.gradle.api.Project;
  * @author <a href="https://github.com/ppalaga">Peter Palaga</a>
  */
 public class EditorconfigGradlePlugin implements Plugin<Project> {
+    public final static String CONFIGURATTION_NAME = "editorconfig";
+    /** The version of {@code org.ec4j.maven:ec4j-linters} keep in sync with the version in {@code build.gradle} */
+    private static final String LINTERS_VERSION = "0.0.7";
 
     /** {@inheritDoc} */
     @Override
     public void apply(Project project) {
         project.getExtensions().create(EditorconfigExtension.NAME, EditorconfigExtension.class);
+        project.getRepositories().add(project.getRepositories().mavenCentral());
+        project.getConfigurations().maybeCreate(CONFIGURATTION_NAME);
+        final DependencyHandler dependencies = project.getDependencies();
+        dependencies.add(CONFIGURATTION_NAME, "org.ec4j.maven:ec4j-linters:" + LINTERS_VERSION);
+
         project.getTasks().create(EditorconfigCheckTask.NAME, EditorconfigCheckTask.class);
         project.getTasks().create(EditorconfigFormatTask.NAME, EditorconfigFormatTask.class);
     }

--- a/src/main/java/org/ec4j/gradle/LinterConfig.java
+++ b/src/main/java/org/ec4j/gradle/LinterConfig.java
@@ -16,15 +16,20 @@
  */
 package org.ec4j.gradle;
 
+import java.io.Serializable;
+
 import org.ec4j.maven.lint.api.Linter;
-import org.ec4j.maven.linters.TextLinter;
 
 /**
  * A configuration of a {@link Linter}.
  *
  * @author <a href="https://github.com/ppalaga">Peter Palaga</a>
  */
-public class LinterConfig {
+public class LinterConfig implements Serializable {
+
+    private static final long serialVersionUID = 3565859468589596550L;
+
+    private static final String LINTERS_PACKAGE = "org.ec4j.maven.linters";
 
     private String className;
     private boolean enabled = true;
@@ -34,7 +39,7 @@ public class LinterConfig {
     private boolean useDefaultIncludesAndExcludes = true;
 
     public String getClassName() {
-        return className.indexOf('.') < 0 ? TextLinter.class.getPackage().getName() + "." + className + "Linter"
+        return className.indexOf('.') < 0 ? LINTERS_PACKAGE + "." + className + "Linter"
                 : className;
     }
 

--- a/src/main/java/org/ec4j/gradle/runtime/EditorconfigInvoker.java
+++ b/src/main/java/org/ec4j/gradle/runtime/EditorconfigInvoker.java
@@ -1,0 +1,185 @@
+/*
+ * Copyright (c) 2018 EditorConfig Gradle Plugin
+ * project contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ec4j.gradle.runtime;
+
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Set;
+
+import javax.inject.Inject;
+
+import org.ec4j.core.Cache.Caches;
+import org.ec4j.core.Resource.Resources;
+import org.ec4j.core.ResourceProperties;
+import org.ec4j.core.ResourcePropertiesService;
+import org.ec4j.core.model.PropertyType;
+import org.ec4j.gradle.CollectingLogger;
+import org.ec4j.gradle.EditorconfigCheckTask;
+import org.ec4j.gradle.EditorconfigFormatTask;
+import org.ec4j.gradle.LinterConfig;
+import org.ec4j.maven.lint.api.EditableResource;
+import org.ec4j.maven.lint.api.FormatException;
+import org.ec4j.maven.lint.api.FormattingHandler;
+import org.ec4j.maven.lint.api.Linter;
+import org.ec4j.maven.lint.api.LinterRegistry;
+import org.ec4j.maven.lint.api.Logger;
+import org.ec4j.maven.lint.api.Logger.LogLevel;
+import org.ec4j.maven.lint.api.Resource;
+import org.ec4j.maven.lint.api.ViolationCollector;
+import org.ec4j.maven.lint.api.ViolationHandler;
+import org.gradle.api.GradleException;
+
+/**
+ * A {@link Runnable} suitable for being invoked in an isolated class loader.
+ *
+ * @author <a href="https://github.com/ppalaga">Peter Palaga</a>
+ */
+public class EditorconfigInvoker implements Runnable {
+
+    interface ResourceFactory {
+        Resource createResource(Path absFile, Path relFile, Charset encoding);
+    }
+
+    public static final String FORMAT_EXCEPTION_PREFIX = FormatException.class.getName() + "\n";
+
+    private static LinterRegistry buildLinterRegistry(boolean isAddLintersFromClassPath, List<LinterConfig> linters,
+            ClassLoader cl, Logger log) {
+        final LinterRegistry.Builder linterRegistryBuilder = LinterRegistry.builder().log(log);
+        if (isAddLintersFromClassPath) {
+            linterRegistryBuilder.scan(cl);
+        }
+        if (linters != null && !linters.isEmpty()) {
+            for (LinterConfig linter : linters) {
+                if (linter.isEnabled()) {
+                    linterRegistryBuilder.entry(linter.getId(), linter.getClassName(), cl, linter.getIncludes(),
+                            linter.getExcludes(), linter.isUseDefaultIncludesAndExcludes());
+                } else {
+                    linterRegistryBuilder.removeEntry(linter.getId());
+                }
+            }
+        }
+
+        return linterRegistryBuilder.build();
+
+    }
+
+    private final Path basedirPath;
+    private final Charset charset;
+    private final boolean failOnNoMatchingProperties;
+    private final ViolationHandler handler;
+    private final Set<String> includedFiles;
+    private final LinterRegistry linterRegistry;
+    private final CollectingLogger log;
+    private final ResourceFactory resourceFactory;
+
+    @Inject
+    public EditorconfigInvoker(String taskClass, Set<String> includedFiles, String basedirPath, String charset,
+            boolean isFailOnFormatViolation, boolean isBackUp, String backupSuffix, boolean isAddLintersFromClassPath,
+            List<LinterConfig> linters, boolean failOnNoMatchingProperties) {
+        super();
+        this.log = new CollectingLogger(LogLevel.TRACE);
+        if (taskClass.startsWith(EditorconfigCheckTask.class.getName())) {
+            this.handler = new ViolationCollector(isFailOnFormatViolation, "./gradlew editorconfigFormat", log);
+            this.resourceFactory = new ResourceFactory() {
+                @Override
+                public Resource createResource(Path absFile, Path relFile, Charset encoding) {
+                    return new Resource(absFile, relFile, encoding);
+                }
+            };
+        } else if (taskClass.startsWith(EditorconfigFormatTask.class.getName())) {
+            this.handler = new FormattingHandler(isBackUp, backupSuffix, log);
+            this.resourceFactory = new ResourceFactory() {
+                @Override
+                public Resource createResource(Path absFile, Path relFile, Charset encoding) {
+                    return new EditableResource(absFile, relFile, encoding);
+                }
+            };
+        } else {
+            throw new IllegalStateException(String.format("Expected %s or %s; got %s",
+                    EditorconfigCheckTask.class.getName(), EditorconfigFormatTask.class.getName(), taskClass));
+        }
+        this.includedFiles = includedFiles;
+        this.basedirPath = Paths.get(basedirPath);
+        this.charset = Charset.forName(charset);
+        final ClassLoader invokerCl = EditorconfigInvoker.class.getClassLoader();
+        this.linterRegistry = buildLinterRegistry(isAddLintersFromClassPath, linters, invokerCl, log);
+        this.failOnNoMatchingProperties = failOnNoMatchingProperties;
+    }
+
+    @Override
+    public void run() {
+        FormatException formatException = null;
+        try {
+
+            final ResourcePropertiesService resourcePropertiesService = ResourcePropertiesService.builder() //
+                    .cache(Caches.permanent()) //
+                    .build();
+            handler.startFiles();
+            boolean propertyMatched = false;
+            for (String includedFile : includedFiles) {
+                final Path absFile = Paths.get(includedFile);
+                final Path file = basedirPath.relativize(absFile);
+                log.info("Processing file '{}'", file);
+                final ResourceProperties editorConfigProperties = resourcePropertiesService
+                        .queryProperties(Resources.ofPath(absFile, charset));
+                if (!editorConfigProperties.getProperties().isEmpty()) {
+                    propertyMatched = true;
+                    final Charset useEncoding = Charset
+                            .forName(editorConfigProperties.getValue(PropertyType.charset, charset.name(), true));
+                    final Resource resource = resourceFactory.createResource(absFile, file, useEncoding);
+                    final List<Linter> filteredLinters = linterRegistry.filter(file);
+                    ViolationHandler.ReturnState state = ViolationHandler.ReturnState.RECHECK;
+                    while (state != ViolationHandler.ReturnState.FINISHED) {
+                        for (Linter linter : filteredLinters) {
+                            if (log.isDebugEnabled()) {
+                                log.debug("Processing file '{}' using linter {}", file, linter.getClass().getName());
+                            }
+                            handler.startFile(resource);
+                            linter.process(resource, editorConfigProperties, handler);
+                        }
+                        state = handler.endFile();
+                    }
+                }
+            }
+            if (!propertyMatched) {
+                if (failOnNoMatchingProperties) {
+                    log.error("No .editorconfig properties applicable for files under '{}'", basedirPath);
+                } else {
+                    log.warn("No .editorconfig properties applicable for files under '{}'", basedirPath);
+                }
+            }
+            handler.endFiles();
+        } catch (IOException e) {
+            throw new GradleException(e.getMessage(), e);
+        } catch (FormatException e) {
+            log.fail(e.getMessage());
+            formatException = e;
+        }
+        final StringBuilder msg = new StringBuilder(FORMAT_EXCEPTION_PREFIX);
+        try {
+            log.serialize(msg);
+        } catch (IOException e1) {
+            throw new RuntimeException(e1);
+        }
+        throw new RuntimeException(msg.toString(), formatException);
+
+    }
+
+}

--- a/src/test/java/org/ec4j/gradle/CollectingLoggerTest.java
+++ b/src/test/java/org/ec4j/gradle/CollectingLoggerTest.java
@@ -1,0 +1,119 @@
+/*
+ * Copyright (c) 2018 EditorConfig Gradle Plugin
+ * project contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ec4j.gradle;
+
+import java.io.IOException;
+import java.util.AbstractMap;
+import java.util.List;
+import java.util.Map.Entry;
+
+import org.ec4j.maven.lint.api.Logger.LogLevel;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class CollectingLoggerTest {
+
+    @Test
+    public void serialize() throws IOException {
+        final CollectingLogger log = new CollectingLogger(LogLevel.TRACE);
+
+        log.trace("trace trace\nstr{}", 1);
+        log.debug("debug debug\rstr{}", 2);
+        log.info("info info\r\nstr{}", 3);
+        log.warn("warn warn str{}", 4);
+        log.error("error error\tstr{}", 5);
+        log.fail("fail\nfail\n");
+
+        final String prefix = "bla bla\n";
+        final StringBuilder out = new StringBuilder();
+        out.append(prefix);
+        log.serialize(out);
+
+        final String serialized = out.toString();
+        Assert.assertEquals(prefix //
+                + "TRACE trace trace\\nstr1\n" //
+                + "DEBUG debug debug\\rstr2\n" //
+                + "INFO info info\\r\\nstr3\n" //
+                + "WARN warn warn str4\n" //
+                + "ERROR error error\tstr5\n" //
+                + "FAILURE fail\\nfail\\n", //
+                serialized);
+
+        final CollectingLogger.LogMessages data = CollectingLogger.deserialize(prefix.length(), serialized);
+        final List<Entry<String, String>> messages = data.getMessages();
+        int i = 0;
+        Assert.assertEquals(5, messages.size());
+        Assert.assertEquals(new AbstractMap.SimpleImmutableEntry<String, String>("TRACE", "trace trace\nstr1"),
+                messages.get(i++));
+        Assert.assertEquals(new AbstractMap.SimpleImmutableEntry<String, String>("DEBUG", "debug debug\rstr2"),
+                messages.get(i++));
+        Assert.assertEquals(new AbstractMap.SimpleImmutableEntry<String, String>("INFO", "info info\r\nstr3"),
+                messages.get(i++));
+        Assert.assertEquals(new AbstractMap.SimpleImmutableEntry<String, String>("WARN", "warn warn str4"),
+                messages.get(i++));
+        Assert.assertEquals(new AbstractMap.SimpleImmutableEntry<String, String>("ERROR", "error error\tstr5"),
+                messages.get(i++));
+
+        final String failureMessage = data.getFailureMessage();
+        Assert.assertEquals("fail\nfail\n", failureMessage);
+
+    }
+
+    @Test
+    public void serializeNoFailure() throws IOException {
+        final CollectingLogger log = new CollectingLogger(LogLevel.TRACE);
+
+        log.trace("trace trace\nstr{}", 1);
+        log.debug("debug debug\rstr{}", 2);
+        log.info("info info\r\nstr{}", 3);
+        log.warn("warn warn str{}", 4);
+        log.error("error error\tstr{}", 5);
+
+        final String prefix = "bla bla\n";
+        final StringBuilder out = new StringBuilder();
+        out.append(prefix);
+        log.serialize(out);
+
+        final String serialized = out.toString();
+        Assert.assertEquals(prefix //
+                + "TRACE trace trace\\nstr1\n" //
+                + "DEBUG debug debug\\rstr2\n" //
+                + "INFO info info\\r\\nstr3\n" //
+                + "WARN warn warn str4\n" //
+                + "ERROR error error\tstr5\n", //
+                serialized);
+
+        final CollectingLogger.LogMessages data = CollectingLogger.deserialize(prefix.length(), serialized);
+        final List<Entry<String, String>> messages = data.getMessages();
+        int i = 0;
+        Assert.assertEquals(5, messages.size());
+        Assert.assertEquals(new AbstractMap.SimpleImmutableEntry<String, String>("TRACE", "trace trace\nstr1"),
+                messages.get(i++));
+        Assert.assertEquals(new AbstractMap.SimpleImmutableEntry<String, String>("DEBUG", "debug debug\rstr2"),
+                messages.get(i++));
+        Assert.assertEquals(new AbstractMap.SimpleImmutableEntry<String, String>("INFO", "info info\r\nstr3"),
+                messages.get(i++));
+        Assert.assertEquals(new AbstractMap.SimpleImmutableEntry<String, String>("WARN", "warn warn str4"),
+                messages.get(i++));
+        Assert.assertEquals(new AbstractMap.SimpleImmutableEntry<String, String>("ERROR", "error error\tstr5"),
+                messages.get(i++));
+
+        final String failureMessage = data.getFailureMessage();
+        Assert.assertNull(failureMessage);
+
+    }
+}

--- a/src/test/java/org/ec4j/gradle/EditorconfigGradlePluginITest.java
+++ b/src/test/java/org/ec4j/gradle/EditorconfigGradlePluginITest.java
@@ -25,7 +25,10 @@ import java.nio.file.Paths;
 
 import org.gradle.testkit.runner.BuildResult;
 import org.gradle.testkit.runner.GradleRunner;
+import org.gradle.testkit.runner.InvalidPluginMetadataException;
+import org.gradle.testkit.runner.InvalidRunnerConfigurationException;
 import org.gradle.testkit.runner.TaskOutcome;
+import org.gradle.testkit.runner.UnexpectedBuildSuccess;
 import org.junit.Assert;
 import org.junit.BeforeClass;
 import org.junit.Test;
@@ -75,59 +78,66 @@ public class EditorconfigGradlePluginITest {
         final String projectName = "defaults";
         Path testProjectPath = init(projectName, "checkDefaults");
 
-        BuildResult result = GradleRunner.create().withProjectDir(testProjectPath.toFile()).withArguments( //
-                EditorconfigCheckTask.NAME //
-                , "--debug" //
-        // , "--stacktrace" //
-        ) //
-                // .withDebug(true)
-                .withPluginClasspath() //
-                .buildAndFail();
+        try {
+            BuildResult result = GradleRunner.create().withProjectDir(testProjectPath.toFile()).withArguments( //
+                    EditorconfigCheckTask.NAME //
+                    , "--debug" //
+                    , "--stacktrace" //
+            ) //
+                    .withDebug(true)
+                    .withPluginClasspath() //
+                    .buildAndFail();
 
-        final String logText = result.getOutput();
-        Files.write(testProjectPath.resolve("log.txt"), logText.getBytes(StandardCharsets.UTF_8));
+            final String logText = result.getOutput();
+            Files.write(testProjectPath.resolve("log.txt"), logText.getBytes(StandardCharsets.UTF_8));
 
-        Assert.assertEquals(result.task(":" + EditorconfigCheckTask.NAME).getOutcome(), TaskOutcome.FAILED);
+            Assert.assertEquals(result.task(":" + EditorconfigCheckTask.NAME).getOutcome(), TaskOutcome.FAILED);
 
-        assertLogText(projectName, logText,
-                "Processing file '.editorconfig' using linter org.ec4j.maven.linters.TextLinter");
-        assertLogText(projectName, logText, "No formatting violations found in file '.editorconfig'");
-        assertLogText(projectName, logText,
-                "Processing file 'build.gradle' using linter org.ec4j.maven.linters.TextLinter");
-        assertLogText(projectName, logText, "No formatting violations found in file 'build.gradle'");
-        assertLogText(projectName, logText,
-                "Processing file 'src/main/java/org/ec4j/maven/it/defaults/App.java' using linter org.ec4j.maven.linters.TextLinter"
-                        .replace('/', File.separatorChar));
-        assertLogText(projectName, logText,
-                "No formatting violations found in file 'src/main/java/org/ec4j/maven/it/defaults/App.java'"
-                        .replace('/', File.separatorChar));
-        assertLogText(projectName, logText,
-                "Processing file 'src/main/resources/trailing-whitespace.txt' using linter org.ec4j.maven.linters.TextLinter"
-                        .replace('/', File.separatorChar));
-        assertLogText(projectName, logText,
-                "src/main/resources/trailing-whitespace.txt@1,7: Delete 2 characters - violates trim_trailing_whitespace = true, reported by org.ec4j.maven.linters.TextLinter"
-                        .replace('/', File.separatorChar));
-        assertLogText(projectName, logText,
-                "Processing file 'src/main/resources/indent.xml' using linter org.ec4j.maven.linters.TextLinter"
-                        .replace('/', File.separatorChar));
-        assertLogText(projectName, logText,
-                "Processing file 'src/main/resources/indent.xml' using linter org.ec4j.maven.linters.XmlLinter"
-                        .replace('/', File.separatorChar));
-        assertLogText(projectName, logText,
-                "src/main/resources/indent.xml@23,5: Delete 1 character - violates indent_style = space, indent_size = 2, reported by org.ec4j.maven.linters.XmlLinter"
-                        .replace('/', File.separatorChar));
-        assertLogText(projectName, logText,
-                "src/main/resources/indent.xml@24,3: Delete 2 characters - violates indent_style = space, indent_size = 2, reported by org.ec4j.maven.linters.XmlLinter"
-                        .replace('/', File.separatorChar));
-        assertLogText(projectName, logText,
-                "Processing file 'README.adoc' using linter org.ec4j.maven.linters.TextLinter");
-        assertLogText(projectName, logText,
-                "README.adoc@2,1: Delete 2 characters - violates trim_trailing_whitespace = true, reported by org.ec4j.maven.linters.TextLinter");
-        assertLogText(projectName, logText, "Checked 6 files");
-        assertLogText(projectName, logText, ":" + EditorconfigCheckTask.NAME + " FAILED");
-        assertLogText(projectName, logText, "There are .editorconfig violations. You may want to run");
-        assertLogText(projectName, logText, "./gradlew editorconfigFormat");
-        assertLogText(projectName, logText, "to fix them automagically.");
+            assertLogText(projectName, logText,
+                    "Processing file '.editorconfig' using linter org.ec4j.maven.linters.TextLinter");
+            assertLogText(projectName, logText, "No formatting violations found in file '.editorconfig'");
+            assertLogText(projectName, logText,
+                    "Processing file 'build.gradle' using linter org.ec4j.maven.linters.TextLinter");
+            assertLogText(projectName, logText, "No formatting violations found in file 'build.gradle'");
+            assertLogText(projectName, logText,
+                    "Processing file 'src/main/java/org/ec4j/maven/it/defaults/App.java' using linter org.ec4j.maven.linters.TextLinter"
+                            .replace('/', File.separatorChar));
+            assertLogText(projectName, logText,
+                    "No formatting violations found in file 'src/main/java/org/ec4j/maven/it/defaults/App.java'"
+                            .replace('/', File.separatorChar));
+            assertLogText(projectName, logText,
+                    "Processing file 'src/main/resources/trailing-whitespace.txt' using linter org.ec4j.maven.linters.TextLinter"
+                            .replace('/', File.separatorChar));
+            assertLogText(projectName, logText,
+                    "src/main/resources/trailing-whitespace.txt@1,7: Delete 2 characters - violates trim_trailing_whitespace = true, reported by org.ec4j.maven.linters.TextLinter"
+                            .replace('/', File.separatorChar));
+            assertLogText(projectName, logText,
+                    "Processing file 'src/main/resources/indent.xml' using linter org.ec4j.maven.linters.TextLinter"
+                            .replace('/', File.separatorChar));
+            assertLogText(projectName, logText,
+                    "Processing file 'src/main/resources/indent.xml' using linter org.ec4j.maven.linters.XmlLinter"
+                            .replace('/', File.separatorChar));
+            assertLogText(projectName, logText,
+                    "src/main/resources/indent.xml@23,5: Delete 1 character - violates indent_style = space, indent_size = 2, reported by org.ec4j.maven.linters.XmlLinter"
+                            .replace('/', File.separatorChar));
+            assertLogText(projectName, logText,
+                    "src/main/resources/indent.xml@24,3: Delete 2 characters - violates indent_style = space, indent_size = 2, reported by org.ec4j.maven.linters.XmlLinter"
+                            .replace('/', File.separatorChar));
+            assertLogText(projectName, logText,
+                    "Processing file 'README.adoc' using linter org.ec4j.maven.linters.TextLinter");
+            assertLogText(projectName, logText,
+                    "README.adoc@2,1: Delete 2 characters - violates trim_trailing_whitespace = true, reported by org.ec4j.maven.linters.TextLinter");
+            assertLogText(projectName, logText, "Checked 6 files");
+            assertLogText(projectName, logText, ":" + EditorconfigCheckTask.NAME + " FAILED");
+            assertLogText(projectName, logText, "There are .editorconfig violations. You may want to run");
+            assertLogText(projectName, logText, "./gradlew editorconfigFormat");
+            assertLogText(projectName, logText, "to fix them automagically.");
+        } catch (UnexpectedBuildSuccess e) {
+            final BuildResult result = e.getBuildResult();
+            final String logText = result.getOutput();
+            Files.write(testProjectPath.resolve("log.txt"), logText.getBytes(StandardCharsets.UTF_8));
+            throw e;
+        }
 
     }
 
@@ -136,47 +146,55 @@ public class EditorconfigGradlePluginITest {
         final String projectName = "extension";
         final Path testProjectPath = init(projectName, "checkExtension");
 
-        BuildResult result = GradleRunner.create().withProjectDir(testProjectPath.toFile()).withArguments( //
-                EditorconfigCheckTask.NAME //
-                , "--debug" //
-        // , "--stacktrace" //
-        ) //
-                // .withDebug(true)
-                .withPluginClasspath() //
-                .buildAndFail();
+        try {
+            BuildResult result = GradleRunner.create().withProjectDir(testProjectPath.toFile()).withArguments( //
+                    EditorconfigCheckTask.NAME //
+                    , "--debug" //
+            // , "--stacktrace" //
+            ) //
+                    // .withDebug(true)
+                    .withPluginClasspath() //
+                    .buildAndFail();
 
-        final String logText = result.getOutput();
-        Files.write(testProjectPath.resolve("log.txt"), logText.getBytes(StandardCharsets.UTF_8));
+            final String logText = result.getOutput();
+            Files.write(testProjectPath.resolve("log.txt"), logText.getBytes(StandardCharsets.UTF_8));
 
-        Assert.assertEquals(result.task(":" + EditorconfigCheckTask.NAME).getOutcome(), TaskOutcome.FAILED);
+            Assert.assertEquals(result.task(":" + EditorconfigCheckTask.NAME).getOutcome(), TaskOutcome.FAILED);
 
 
-        assertNoLogText(projectName, logText,
-                "Processing file '.editorconfig' using linter org.ec4j.maven.linters.TextLinter");
-        assertNoLogText(projectName, logText,
-                "Processing file 'build.gradle' using linter org.ec4j.maven.linters.TextLinter");
-        assertLogText(projectName, logText,
-                "Processing file 'src/main/java/org/ec4j/maven/it/defaults/App.java' using linter org.ec4j.maven.linters.TextLinter"
-                        .replace('/', File.separatorChar));
-        assertLogText(projectName, logText,
-                "No formatting violations found in file 'src/main/java/org/ec4j/maven/it/defaults/App.java'"
-                        .replace('/', File.separatorChar));
-        assertLogText(projectName, logText,
-                "Processing file 'src/main/resources/trailing-whitespace.txt' using linter org.ec4j.maven.linters.TextLinter"
-                        .replace('/', File.separatorChar));
-        assertLogText(projectName, logText,
-                "src/main/resources/trailing-whitespace.txt@1,7: Delete 2 characters - violates trim_trailing_whitespace = true, reported by org.ec4j.maven.linters.TextLinter"
-                        .replace('/', File.separatorChar));
-        assertNoLogText(projectName, logText,
-                "Processing file 'src/main/resources/indent.xml' using linter org.ec4j.maven.linters.TextLinter"
-                        .replace('/', File.separatorChar));
-        assertNoLogText(projectName, logText,
-                "Processing file 'README.adoc' using linter org.ec4j.maven.linters.TextLinter");
-        assertLogText(projectName, logText, "Checked 2 files");
-        assertLogText(projectName, logText, ":" + EditorconfigCheckTask.NAME + " FAILED");
-        assertLogText(projectName, logText, "There are .editorconfig violations. You may want to run");
-        assertLogText(projectName, logText, "./gradlew editorconfigFormat");
-        assertLogText(projectName, logText, "to fix them automagically.");
+            assertNoLogText(projectName, logText,
+                    "Processing file '.editorconfig' using linter org.ec4j.maven.linters.TextLinter");
+            assertNoLogText(projectName, logText,
+                    "Processing file 'build.gradle' using linter org.ec4j.maven.linters.TextLinter");
+            assertLogText(projectName, logText,
+                    "Processing file 'src/main/java/org/ec4j/maven/it/defaults/App.java' using linter org.ec4j.maven.linters.TextLinter"
+                            .replace('/', File.separatorChar));
+            assertLogText(projectName, logText,
+                    "No formatting violations found in file 'src/main/java/org/ec4j/maven/it/defaults/App.java'"
+                            .replace('/', File.separatorChar));
+            assertLogText(projectName, logText,
+                    "Processing file 'src/main/resources/trailing-whitespace.txt' using linter org.ec4j.maven.linters.TextLinter"
+                            .replace('/', File.separatorChar));
+            assertLogText(projectName, logText,
+                    "src/main/resources/trailing-whitespace.txt@1,7: Delete 2 characters - violates trim_trailing_whitespace = true, reported by org.ec4j.maven.linters.TextLinter"
+                            .replace('/', File.separatorChar));
+            assertNoLogText(projectName, logText,
+                    "Processing file 'src/main/resources/indent.xml' using linter org.ec4j.maven.linters.TextLinter"
+                            .replace('/', File.separatorChar));
+            assertNoLogText(projectName, logText,
+                    "Processing file 'README.adoc' using linter org.ec4j.maven.linters.TextLinter");
+            assertLogText(projectName, logText, "Checked 2 files");
+            assertLogText(projectName, logText, ":" + EditorconfigCheckTask.NAME + " FAILED");
+            assertLogText(projectName, logText, "There are .editorconfig violations. You may want to run");
+            assertLogText(projectName, logText, "./gradlew editorconfigFormat");
+            assertLogText(projectName, logText, "to fix them automagically.");
+        } catch (UnexpectedBuildSuccess e) {
+            final BuildResult result = e.getBuildResult();
+            final String logText = result.getOutput();
+            Files.write(testProjectPath.resolve("log.txt"), logText.getBytes(StandardCharsets.UTF_8));
+            throw e;
+        }
+
 
     }
 

--- a/src/test/projects/defaults-formatted/.editorconfig
+++ b/src/test/projects/defaults-formatted/.editorconfig
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2017 EditorConfig Maven Plugin
+# Copyright (c) 2018 EditorConfig Gradle Plugin
 # project contributors as indicated by the @author tags.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");

--- a/src/test/projects/defaults/.editorconfig
+++ b/src/test/projects/defaults/.editorconfig
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2017 EditorConfig Maven Plugin
+# Copyright (c) 2018 EditorConfig Gradle Plugin
 # project contributors as indicated by the @author tags.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");

--- a/src/test/projects/extension/.editorconfig
+++ b/src/test/projects/extension/.editorconfig
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2017 EditorConfig Maven Plugin
+# Copyright (c) 2018 EditorConfig Gradle Plugin
 # project contributors as indicated by the @author tags.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");

